third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -236,6 +247,123 @@ impl<'a> FunctionGenerator<'a> {
+            changed = false;
+            let mut this_index = 0;
+            while this_index < (self.code.len() - 1) {
+                if deleted[this_index] {
+                    this_index += 1;
+                    continue;
+                }
+                let this_bc = &self.code[this_index];
+                if let Some(mut next_index) = deleted[(this_index + 1)..].iter().position(|v| !*v) {
+                    next_index += this_index + 1;
+                    // Note that we should not delete a sequence of pairs of instructions if we can jump to
+                    // any of them (except the first one).
+                    if is_jump_target[(this_index + 1)..=next_index]
+                        .iter()
+                        .all(|v| !*v)
+                    {
+                        let next_bc = &self.code[next_index];
+                        use FF::Bytecode::*;
+                        // Each of these pairs of instructions are equivalent to a no-op, so they can be safely
+                        // removed assuming the above conditions are met.
+                        match (this_bc, next_bc) {
+                            (StLoc(u), MoveLoc(v))
+                            | (CopyLoc(u), StLoc(v))
+                            | (MoveLoc(u), StLoc(v))
+                                if *u == *v =>
+                            {
+                                deleted[this_index] = true;
+                                deleted[next_index] = true;
+                                no_deletions = false;
+                                changed = true;
+                            },
+                            _ => {},
+                        }
+                    }
+                    this_index = next_index;
+                } else {
+                    break; // no more instructions left for the inner loop
+                }
+            }
+        }
+        if no_deletions {
+            return;
+        }
+        // Remove deleted instructions.
+        let mut new_code = vec![];
+        // Because of instructions being deleted, their offsets may have changed.
+        let mut offset_remap: Vec<CodeOffset> =
+            (0..self.code.len()).map(|x| x as CodeOffset).collect();
+        for (i, bc) in self.code.iter().enumerate() {
+            if !deleted[i] {
+                offset_remap[i] = new_code.len() as CodeOffset;
+                new_code.push(bc.clone());
+            }
+        }
+        for i in 0..self.code.len() {
+            if deleted[i] {
+                if let Some(next_undeleted) = deleted[(i + 1)..].iter().position(|v| !*v) {
+                    offset_remap[i] = offset_remap[i + 1 + next_undeleted];
+                } else {
+                    unreachable!("There should be at least one undeleted instruction after every deleted one")
+                }
+            }
+        }
+        self.code = new_code;
+        // Update label info withe new code offsets.
```

Comment:
How about a separate function for each of these sections?

Reply:
Addressed by significantly rewriting to improve readability and maintainability.